client/my-sites/controller.js

```diff
@@ -383,12 +384,12 @@ export function siteSelection( context, next ) {
 	} else {
 		const selectOnSitesChange = () => {
 			const freshSiteId = getSiteId( getState(), siteFragment );
-			dispatch( setSelectedSiteId( freshSiteId ) );
 
 			if ( getSite( getState(), freshSiteId ) ) {
+				dispatch( setSelectedSiteId( freshSiteId ) );
 				onSelectedSiteAvailable( context );
-			} else {
-				// if sites have loaded, but siteId is invalid, redirect to allSitesPath
+			} else if ( hasAllSitesList( getState() ) ) {
+				// If all sites have loaded, but siteId is still invalid, redirect to allSitesPath.
 				page.redirect( allSitesPath );
```

Comment:
Previously, selected site was set to `null` before redirecting, now `setSelectedSiteId` is not called at all and `state.ui.selectedSiteId` keeps the previous value. Maybe it's not a problem, maybe it is, I'm not sure. Can you verify that?

Other than that, the patch looks good.

Reply:
Yes, good point. After the redirect is called, `siteSelection` is called again (for the redirected route). This time, there's no `siteFragment` as it's an all sites view route. So different condition is triggered in `siteSelection`: https://github.com/Automattic/wp-calypso/blob/b7d6850adc2a48b3547ec0dfd45b30a1b46ed0ac/client/my-sites/controller.js#L370 It's set to `null` there.